File: ruby/robot-name/robot_name.rb
class Robot

  attr_reader :name

  def name
    @name ||= "#{letters}#{numbers}"
  end

  def letters
    ('A'..'Z').to_a.sample(2).join
  end

  def numbers
    rand(100..999)
  end

  def reset
    @name = nil
  end

end

File: javascript/robot-name/robot-name.js
function Robot() {
  this.name  = getRobotName();
}

Robot.prototype.reset = function() {
  this.name = getRobotName();
  return this;
}

function getRobotName() {
  return randomLetters() + (Math.floor(Math.random() * (999 - 100)) + 100);
}

function randomLetters() {
  var alphabet = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');

    var m = alphabet.length, t, i;

    while (m) {
      i = Math.floor(Math.random() * m--);

      t = alphabet[m];
      alphabet[m] = alphabet[i];
      alphabet[i] = t;
    }
  return alphabet.slice(0, 2).join('');
}

module.exports = Robot;
